### PR TITLE
[testing] introduce the Cluster struct & integration test for causal completion

### PR DESCRIPTION
**Pull request by akichidis**
This PR is introducing:
* The `Cluster` struct to allow us easily bring up a cluster of primary nodes (up to 4 for now) , stop single nodes, start nodes independently (so we can simulate late comers) etc.
* Allow confirm the nodes internals by using the prometheus Registries (setup individual ones by each node)
* An integration test that showcases the issue of the causal completion and the `Failed to process our own vote` error that has been presented on the https://github.com/MystenLabs/narwhal/pull/486 as well.

### CODE REVIEW

File: node/src/execution_state.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
```
Comment (akichidis) [inline node/src/execution_state.rs:1]: Just moved this code out of the `main` file

File: primary/src/metrics.rs [hunk 2]
```diff
@@ -16,35 +16,23 @@ pub(crate) struct Metrics {
     pub(crate) node_metrics: Option<PrimaryMetrics>,
 }
 
-static mut METRICS: Metrics = Metrics {
-    endpoint_metrics: None,
-    primary_endpoint_metrics: None,
-    node_metrics: None,
-};
-static INIT: Once = Once::new();
```
Comment (akichidis) [inline primary/src/metrics.rs:24]: No need to keep the metrics behind a `Once` - since we can pass a Registry we can always pass a separate Registry per test so we don't bump into conflict issues on the default registry

File: primary/tests/causal_completion_tests.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    let mut cluster = Cluster::new(None);
+
+    // start the cluster
+    let nodes = cluster.start(4).await;
+
+    // Let primaries advance little bit
+    tokio::time::sleep(Duration::from_secs(10)).await;
+
+    // Ensure all nodes advanced
+    for node in nodes {
```
Comment (akichidis) [inline primary/tests/causal_completion_tests.rs:26]: I believe at the moment the metrics is the most powerful tool we have to understand what's going on with a node since we are lacking of any endpoints - at least to test the cases that have to do with node advancement, late joiners etc

File: test_utils/src/cluster.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        }
+
+        let mut k = keys(None);
+        let keypair = k.remove(id);
+        let name = keypair.public().clone();
+
+        let registry = primary_metrics_registry(name.clone());
+
+        // Make the data store.
+        let store: NodeStorage<Ed25519PublicKey> = NodeStorage::reopen(temp_dir());
```
[the remark this answers is not in the document]
Comment (laura-makdah) [reply test_utils/src/cluster.rs:86]: That is precisely the functionality we need for testing loading consensus from disk.

File: primary/tests/causal_completion_tests.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        let metric_family = node.registry.gather();
+
+        for metric in metric_family {
```
Comment (laura-makdah) [inline primary/tests/causal_completion_tests.rs:29]: Before this point, don't we need to add certificates to the nodes for processing? Or rather, I think it would be helpful for testing more specific scenarios, but don't quite know how to do that.

File: test_utils/src/cluster.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (akichidis) [reply test_utils/src/cluster.rs:86]: I have opened a follow up PR that allow us to preserve the disk between node restarts (reading from the same RocksDB folder) as currently we always create a new one we a node starts. I believe that would be useful for your tests https://github.com/MystenLabs/narwhal/pull/492

**Review by huitseeker — COMMENTED** [review]
This seems like a good start! Two questions:
- we're reusing the `tracing_subscriber` manually, as we did before in this repo. Is this a good time to upgrade to [telemetry-subscribers](https://github.com/MystenLabs/mysten-infra/tree/main/crates/telemetry-subscribers) as Sui does, hopefully minimizing some maintenance burden? e.g. I can definitely see myself running those tests under tokio-console 
- the absence of a worker seems like a severe limitation given some workflows wait on worker interaction, how do we plan to address this component?

File: primary/tests/causal_completion_tests.rs (hunk 5 above)
Comment (huitseeker) [reply primary/tests/causal_completion_tests.rs:29]: The test introduced here may include useful scaffolding: https://github.com/MystenLabs/narwhal/pull/486

File: test_utils/src/cluster.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+        // The channel returning the result for each transaction's execution.
+        let (tx_transaction_confirmation, _) = channel(Node::CHANNEL_CAPACITY);
+
+        let h = Node::spawn_primary(
```
[the remark this answers is not in the document]
Comment (akichidis) [reply test_utils/src/cluster.rs:91]: On follow up PR I'll spin up worker(s) for each authority - they'll communicate via network as expected. The current test doesn't really need workers as we are testing against empty headers (they don't have payloads) where no contact is required to workers at any point. Very precisely as you said this will be an issue when we make tests where we pump in transactions - that's why I'll amend to spin up worker nodes.

**Comment by akichidis** [comment]
Thanks a lot for the feedback @huitseeker .
> * we're reusing the `tracing_subscriber` manually, as we did before in this repo. Is this a good time to upgrade to [telemetry-subscribers](https://github.com/MystenLabs/mysten-infra/tree/main/crates/telemetry-subscribers) as Sui does, hopefully minimizing some maintenance burden? e.g. I can definitely see myself running those tests under tokio-console

ah thanks for that. Let me have a quick look and see how this could be incorporated in this PR. Otherwise I'll address on a follow up one.

> * the absence of a worker seems like a severe limitation given some workflows wait on worker interaction, how do we plan to address this component?

I've replied on the related comment. I'll include that on follow up PR.
